apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -0,0 +1,626 @@
+            });
+
+        /// <summary>
+        /// Publish a message to the topic specified in <see cref="TopicName"/>.
+        /// </summary>
+        /// <param name="message">The <see cref="PubsubMessage"/> to publish.</param>
+        /// <returns>A task which completes once the message has been published.
+        /// The task <see cref="Task{String}.Result"/> contains the message ID.</returns>
+        public virtual Task<string> PublishAsync(IMessage message) =>
+            PublishAsync(new PubsubMessage
+            {
+                Data = message.ToByteString()
+            });
+
+        /// <summary>
+        /// Publish a message to the topic specified in <see cref="TopicName"/>.
+        /// </summary>
+        /// <param name="message">The <see cref="PubsubMessage"/> to publish.</param>
+        /// <returns>A task which completes once the message has been published.
+        /// The task <see cref="Task{String}.Result"/> contains the message ID.</returns>
+        public virtual Task<string> PublishAsync(ByteString message) =>
+            PublishAsync(new PubsubMessage
+            {
+                Data = message
+            });
+
+        /// <summary>
+        /// Publish a message to the topic specified in <see cref="TopicName"/>.
+        /// </summary>
+        /// <param name="message">The <see cref="PubsubMessage"/> to publish.</param>
+        /// <returns>A task which completes once the message has been published.
+        /// The task <see cref="Task{String}.Result"/> contains the message ID.</returns>
+        public virtual Task<string> PublishAsync(byte[] message) =>
+            PublishAsync(new PubsubMessage
+            {
+                Data = ByteString.CopyFrom(message)
+            });
+
+        /// <summary>
+        /// Retrieve a snapshot of the flow state.
+        /// </summary>
+        /// <returns></returns>
+        public virtual FlowState GetCurrentFlowState() => throw new NotImplementedException();
+
+        /// <summary>
+        /// Shutdown this <see cref="HiPerfPublisher"/>.
+        /// The returned <see cref="Task"/> completes when all queued messages have been published.
+        /// The returned <see cref="Task"/> cancels if the passed <see cref="CancellationToken"/> is cancelled.
+        /// Cancelling the passed <see cref="CancellationToken"/> aborts publishing, and cancels the returned
+        /// <see cref="Task"/> as quickly as possible.
+        /// </summary>
+        /// <param name="hardStopToken">Cancel this <see cref="CancellationToken"/> to abort publishing queued messages.</param>
+        /// <returns>A <see cref="Task"/> that completes when all queued messages have been published; or cancels if the
+        /// passed <see cref="CancellationToken"/> is cancelled.</returns>
+        public virtual Task ShutdownAsync(CancellationToken hardStopToken) => throw new NotImplementedException();
```

Comment:
Interesting - this is quite a different use of a cancellation token. I simultaneously like it as a clean design, and dislike it for being surprising. Good call on not calling the parameter cancellationToken though.